### PR TITLE
fix: Incorrect LEFT JOIN evaluation result on OR conditions

### LINKED ISSUE #10881: Incorrect LEFT JOIN evaluation result on OR conditions
### Describe the bug

When the matching condition is on columns and combined with `OR`, the result retuned by `LEFT JOIN` is similar to `INNER JOIN`, which is incorrect.

`LEFT JOIN` should not reduce the number of rows returned, i.e. it should return all the records from the left table, and the matched records from the right table. When there is no match, the result is a `NULL`, i.e. display as empty.

`INNER JOIN` reduces the number of rows, i.e. it only returns the records that have matching values in both tables. 

### To Reproduce

1. Use DataFusion CLI:
```shell
# in dir datafusion/datafusion-cli
$ cargo build
$ ./target/debug/datafusion-cli
```

2. Write sample data
```sql
CREATE OR REPLACE TABLE employees(emp_id INT, name VARCHAR) AS VALUES (1, 'Alice'), (2, 'Bob'), (3, 'Carol');
CREATE OR REPLACE TABLE department(emp_id INT, department VARCHAR) AS VALUES (1, 'HR'), (3, 'Engineering'), (4, 'Sales');
```
3. This is how `employees` table and `department` table look like
```sql
> SELECT * FROM employees;
+--------+-------+
| emp_id | name  |
+--------+-------+
| 1      | Alice |
| 2      | Bob   |
| 3      | Carol |
+--------+-------+
3 row(s) fetched.
Elapsed 0.009 seconds.

> SELECT * FROM department;
+--------+-------------+
| emp_id | department  |
+--------+-------------+
| 1      | HR          |
| 3      | Engineering |
| 4      | Sales       |
+--------+-------------+
3 row(s) fetched.
Elapsed 0.006 seconds.
```

4. Query
```sql
> SELECT e.emp_id, e.name, d.department
FROM employees e
LEFT JOIN department d
ON (e.name = 'Alice' OR e.name = 'Bob');
+--------+-------+-------------+
| emp_id | name  | department  |
+--------+-------+-------------+
| 1      | Alice | HR          |
| 1      | Alice | Engineering |
| 1      | Alice | Sales       |
| 2      | Bob   | HR          |
| 2      | Bob   | Engineering |
| 2      | Bob   | Sales       |
+--------+-------+-------------+ <-- should have one more row for Carol
6 row(s) fetched.
Elapsed 0.013 seconds.

> SELECT e.emp_id, e.name, d.department
FROM employees e
LEFT JOIN department d
ON (e.name = 'NotExist1' OR e.name = 'NotExist2');
+--------+------+------------+
| emp_id | name | department |
+--------+------+------------+
+--------+------+------------+ <-- should have three rows, 1 row for Alice, 1 row for Bob, and 1 row for Carol
0 row(s) fetched.
Elapsed 0.014 seconds.

> SELECT e.emp_id, e.name, d.department
FROM employees e
LEFT JOIN department d
ON (e.name = 'Alice' OR e.name = 'NotExist');
+--------+-------+-------------+
| emp_id | name  | department  |
+--------+-------+-------------+
| 1      | Alice | HR          |
| 1      | Alice | Engineering |
| 1      | Alice | Sales       |
+--------+-------+-------------+ <-- should have two more rows, 1 row for Bob and 1 row for Carol
3 row(s) fetched.
Elapsed 0.014 seconds.
```

### Expected behavior

Postgres shows the expected results. 

1. In psql cli, drop tables and create new tables with data
```sql
DROP TABLE IF EXISTS employees;
DROP TABLE IF EXISTS department;

# create tables
CREATE TABLE employees(emp_id INT, name VARCHAR);
CREATE TABLE department(emp_id INT, dept_name VARCHAR);

# write sample data
INSERT INTO employees (emp_id, name) VALUES (1, 'Alice'), (2, 'Bob'), (3, 'Carol');
INSERT INTO department (emp_id, dept_name) VALUES (1, 'HR'), (3, 'Engineering'), (4, 'Sales');
```

2. This is how two tables look like:
```sql
chunchun=# SELECT * FROM employees;
 emp_id | name
--------+-------
      1 | Alice
      2 | Bob
      3 | Carol
(3 rows)

chunchun=# SELECT * FROM department;
 emp_id |  dept_name
--------+-------------
      1 | HR
      3 | Engineering
      4 | Sales
(3 rows)
```

3. Query
```sql
SELECT e.emp_id, e.name, d.department
FROM employees AS e
LEFT JOIN department AS d
ON (e.name = 'Alice' OR e.name = 'Bob');
 emp_id | name  | department
--------+-------+-------------
      1 | Alice | HR
      1 | Alice | Engineering
      1 | Alice | Sales
      2 | Bob   | HR
      2 | Bob   | Engineering
      2 | Bob   | Sales
      3 | Carol |
(7 rows)

SELECT e.emp_id, e.name, d.department
FROM employees e
LEFT JOIN department d
ON (e.name = 'NotExist1' OR e.name = 'NotExist2');
 emp_id | name  | department
--------+-------+------------
      1 | Alice |
      2 | Bob   |
      3 | Carol |
(3 rows)

SELECT e.emp_id, e.name, d.department
FROM employees e
LEFT JOIN department d
ON (e.name = 'Alice' OR e.name = 'NotExist');
 emp_id | name  | department
--------+-------+-------------
      1 | Alice | HR
      1 | Alice | Engineering
      1 | Alice | Sales
      2 | Bob   |
      3 | Carol |
(5 rows)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #10881.

### CODE REVIEW

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -441,11 +442,11 @@ fn push_down_all_join(
 
     // Extract from OR clause, generate new predicates for both side of join if possible.
     // We only track the unpushable predicates above.
-    if left_preserved {
+    if on_left_preserved {
```
Comment (viirya) [inline datafusion/optimizer/src/push_down_filter.rs:445]: `keep_predicates`, `join_conditions` are join filter but not predicates from top Filter, we should use `on_lr_is_preserved` instead of `lr_is_preserved`.

**Review by alamb — APPROVED** [review]
Thank you @viirya  (and everyone else who reviewed and commented on this bug)

I went through the code carefully and I pushed a few more tests to this branch. I think it looks great and makes sense to me. 

I will also make a new PR to tweak some comments as well

File: datafusion/sqllogictest/test_files/join.slt [hunk 2]
```diff
@@ -793,3 +793,148 @@ DROP TABLE companies
...
+LEFT JOIN department d
+ON (e.name = 'Alice' OR e.name = 'NotExist');
+----
+1 Alice HR
+1 Alice Engineering
+1 Alice Sales
+2 Bob NULL
+3 Carol NULL
+
+# Can push the ON filter below the JOIN for INNER JOIN (expect to see a filter below the join)
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/join.slt:900]: I added a test that shows the ON clause is pushed down for INNER joins

File: datafusion/sqllogictest/test_files/join.slt [hunk 3]
```diff
@@ -839,6 +845,36 @@ ON (e.name = 'Alice' OR e.name = 'Bob');
 2 Bob Sales
 3 Carol NULL
 
+# neither RIGHT OUTER JOIN
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/join.slt:848]: I also added coverage for `RIGHT` and `FULL` joins

**Comment by alamb** [comment]
The CI error appears to be https://github.com/apache/datafusion/issues/11210 unrelated to this PR

File: datafusion/optimizer/src/push_down_filter.rs [hunk 4]
```diff
@@ -441,11 +442,11 @@ fn push_down_all_join(
...
         left_push.extend(extract_or_clauses_for_join(&keep_predicates, left_schema));
         left_push.extend(extract_or_clauses_for_join(&join_conditions, left_schema));
     }
-    if right_preserved {
+    if on_right_preserved {
```
Comment (mustafasrepo) [inline datafusion/optimizer/src/push_down_filter.rs:449]: We have discussed the conditions, filter predicates and on expression can be pushed down with @ozankabak .
According to us, correct handling seems to be 
```
    if left_preserved {
        left_push.extend(extract_or_clauses_for_join(&keep_predicates, left_schema));
    }
    if on_left_preserved {
        left_push.extend(extract_or_clauses_for_join(&join_conditions, left_schema));
    }
    if right_preserved {
        right_push.extend(extract_or_clauses_for_join(&keep_predicates, right_schema));
    }
    if on_right_preserved {
        right_push.extend(extract_or_clauses_for_join(&join_conditions, right_schema));
    }
```
where, for predicates and join conditions flag used is different. What do you think?

Comment (alamb) [reply datafusion/optimizer/src/push_down_filter.rs:449]: I am not sure I understand what is different about your suggestion.  

My understanding of the code is that to breaks the predicates into:
* `keep_predicates` predicates that must be left as a filter (above the join)
* `join_conditions` predicates that must be left in the join (e.g. `ON` clause)
* `left_push` predicates that can be pushed down to the left input
* `right_push` predicates that can be pushed down to the right input

I think the code, as written, does push predicates from the ON clause down correctly when possible. 

So in other words, I am not sure what changes you are suggesting

Comment (ozankabak) [reply datafusion/optimizer/src/push_down_filter.rs:449]: We could be missing something but it seems like `keep_predicates` and `join_conditions` should be using `preserved` vs. `on_preserved` (respectively) since the former applies as a post-join filter.

File: datafusion/sqllogictest/test_files/join.slt (hunk 2 above)
Comment (viirya) [reply datafusion/sqllogictest/test_files/join.slt:900]: Thanks.

File: datafusion/optimizer/src/push_down_filter.rs (hunk 4 above)
Comment (mustafasrepo) [reply datafusion/optimizer/src/push_down_filter.rs:449]: For left join, `left_preserved=true` and `on_left_preserved=false`.
In current version of the code, nothing from `keep_predicates` and `join_conditions` can be inserted to the `left_push` since `on_left_preserved` flag is `false`. However,  if we use the flag `left_preserved` to decide whether we can insert  something from `keep_predicates`, into `left_push`. We could have inserted additional conditions to the `left_push`. In this sense, current code works correctly, however behaves suboptimal in some cases.

Comment (viirya) [reply datafusion/optimizer/src/push_down_filter.rs]: Actually we wrongly mix join filter predicates and filter predicates into `join_conditions`. I separate them now so only `on_filter_join_conditions` will be considered with `on_preserved`.

File: datafusion/optimizer/src/push_down_filter.rs [hunk 5]
```diff
@@ -435,22 +436,37 @@ fn push_down_all_join(
...
 
     // Extract from OR clause, generate new predicates for both side of join if possible.
     // We only track the unpushable predicates above.
-    if on_left_preserved {
+    if left_preserved {
         left_push.extend(extract_or_clauses_for_join(&keep_predicates, left_schema));
         left_push.extend(extract_or_clauses_for_join(&join_conditions, left_schema));
     }
-    if on_right_preserved {
+    if left_preserved {
```
Comment (ozankabak) [inline datafusion/optimizer/src/push_down_filter.rs:450]: I thınk you meant `right_preserved` here.

File: datafusion/optimizer/src/push_down_filter.rs (hunk 4 above)
Comment (alamb) [reply datafusion/optimizer/src/push_down_filter.rs]: Thank you for the clarifications


> In current version of the code, nothing from keep_predicates and join_conditions can be inserted to the left_push since on_left_preserved flag is false

I agree, but I think this is required for correctness. At least for `join_conditions` it is important not to push them down to the preserved side 

For example, in this query  I am pretty sure it is not correct to push the predicate on `A.a` below the join (the intuition is that all rows from `A` have to have at least one row out of the join). I believe this is what this PR fixes. 

```SQL
SELECT .. FROM A LEFT JOIN B ON (A.a > 5)
```

I think *is* correct to push predicates on B below (as any non matching rows would have been filtered in the join anyways)

```SQL
SELECT .. FROM A LEFT JOIN B ON (B.b > 5)
```

> We could have inserted additional conditions to the left_push. In this sense, current code works correctly, however behaves suboptimal in some cases.

Can you help me write an example? I can add it to this PR and then we can handle improving the case in a future PR.

**Comment by alamb** [comment]
Looks like my comment https://github.com/apache/datafusion/pull/11203#discussion_r1662588836 was in progress when @viirya  changed the code in https://github.com/apache/datafusion/pull/11203/commits/17b2d431e44c07fdb6ffc3f4c81c3add0ba0783a 

Can someone help me understand what additional case this would catch (and I'll go and write some tests that cover it)?

**Review by ozankabak — COMMENTED** [review]
I think you made a typo in your last commit (left an inline comment for it), but other than that I think it is now logically equivalent to @mustafasrepo's suggestion on https://github.com/apache/datafusion/pull/11203#discussion_r1662401407 (although his code seems to use a single `join_conditions` vector and seems more concise).

File: datafusion/optimizer/src/push_down_filter.rs (hunk 5 above)
Comment (viirya) [reply datafusion/optimizer/src/push_down_filter.rs:450]: Oops

File: datafusion/optimizer/src/push_down_filter.rs (hunk 4 above)
Comment (viirya) [reply datafusion/optimizer/src/push_down_filter.rs]: I tried to add a test case with it.

Comment (ozankabak) [reply datafusion/optimizer/src/push_down_filter.rs]: I think we found one example -- @mustafasrepo will post shortly

**Comment by mustafasrepo** [comment]
> Looks like my comment [#11203 (comment)](https://github.com/apache/datafusion/pull/11203#discussion_r1662588836) was in progress when @viirya changed the code in [17b2d43](https://github.com/apache/datafusion/commit/17b2d431e44c07fdb6ffc3f4c81c3add0ba0783a)
> 
> Can someone help me understand what additional case this would catch (and I'll go and write some tests that cover it)?

I have constructed an example that triggers this test case (query is a bit weird. I have constructed it to trigger `extract_or_clauses_for_join` call):
```
explain SELECT e.emp_id, e.name, d.dept_name
FROM employees AS e
LEFT JOIN department AS d
ON e.emp_id = d.emp_id
WHERE ((dept_name != 'Engineering' AND e.name = 'Alice') OR (name != 'Alice' AND e.name = 'Carol'));
```
this query generates following plan with recent changes
```
+   logical_plan
+   01)Filter: d.dept_name != Utf8("Engineering") AND e.name = Utf8("Alice") OR e.name != Utf8("Alice") AND e.name = Utf8("Carol")
+   02)--Projection: e.emp_id, e.name, d.dept_name
+   03)----Left Join: e.emp_id = d.emp_id
+   04)------SubqueryAlias: e
+   05)--------Filter: employees.name = Utf8("Alice") OR employees.name != Utf8("Alice") AND employees.name = Utf8("Carol")
+   06)----------TableScan: employees projection=[emp_id, name]
+   07)------SubqueryAlias: d
+   08)--------TableScan: department projection=[emp_id, dept_name]
```
previously it was generating
```
+   logical_plan
+   01)Filter: d.dept_name != Utf8("Engineering") AND e.name = Utf8("Alice") OR e.name != Utf8("Alice") AND e.name = Utf8("Carol")
+   02)--Projection: e.emp_id, e.name, d.dept_name
+   03)----Left Join: e.emp_id = d.emp_id
+   04)------SubqueryAlias: e
+   05)--------TableScan: employees projection=[emp_id, name]
+   06)------SubqueryAlias: d
+   07)--------TableScan: department projection=[emp_id, dept_name]
```
where both are correct. However, first plan is a bit better

**Comment by viirya** [comment]
> I think you made a typo in your last commit (left an inline comment for it), but other than that I think it is now logically equivalent to @mustafasrepo's suggestion on [#11203 (comment)](https://github.com/apache/datafusion/pull/11203#discussion_r1662401407) (although his code seems to use a single `join_conditions` vector and seems more concise).

Yea, I think it is correct. I made the change to make it more obvious to more people. Because they need to figure out that `join_conditions` can only contains parent Filter predicates for inner join and for inner join, both sides are preserved (either `preserved` or `on_preserved`). So for `join_conditions`, we can simply use `on_preserved`.

I can use the suggestion if you think it is better.

**Comment by alamb** [comment]
> where both are correct. However, first plan is a bit better

Thank you @mustafasrepo  - that makes sense to me. Let's also add that example to the tests (I can do so in a follow on PR or if we are going to change this PR again we can do so)

**Comment by mustafasrepo** [comment]
> I can use the suggestion if you think it is better.

As long as we have the functionality, both versions are fine.

**Comment by viirya** [comment]
Thanks @mustafasrepo @ozankabak @alamb . I added the test case.

**Review by ozankabak — APPROVED** [review]
Great collaboration!

**Comment by ozankabak** [comment]
Should we go ahead and merge? Are we waiting for more feedback?

**Comment by alamb** [comment]
🚀  I agree it was time to merge

Thanks again everyone
